File: web/lib/api/character.ts
import { apiFetch } from '@/lib/http/client'

export interface CharacterItem {
  code: string
  name: string
  category: 'hair' | 'hat' | 'outfit' | 'aura'
  sprite_url: string
  rarity: 'common' | 'rare' | 'epic' | 'legendary'
  required_level: number
  required_title_code?: string
  is_unlocked: boolean
}

export interface UserCharacter {
  user_id: string
  equipped_hair?: string
  equipped_hat?: string
  equipped_outfit?: string
  equipped_aura?: string
  updated_at: string
}

export interface CharacterSummary {
  character: UserCharacter
  inventory: CharacterItem[]
}

export function getCharacterSummary() {
  return apiFetch<{ data: CharacterSummary }>('/api/v1/character')
}

export function equipCharacterItem(itemCode: string) {
  return apiFetch<{ data: { character: UserCharacter } }>('/api/v1/character/equip', {
    method: 'POST',
    body: JSON.stringify({ item_code: itemCode }),
  })
}

File: common-api/internal/model/character.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type CharacterItem struct {
	Code              string    `json:"code"`
	Name              string    `json:"name"`
	Category          string    `json:"category"`
	SpriteURL         string    `json:"sprite_url"`
	Rarity            string    `json:"rarity"`
	RequiredLevel     int       `json:"required_level"`
	RequiredTitleCode *string   `json:"required_title_code,omitempty"`
	PriceGold         int       `json:"price_gold"`
	IsShopItem        bool      `json:"is_shop_item"`
	IsOwned           bool      `json:"is_owned"`
	IsLevelUnlocked   bool      `json:"is_level_unlocked"`
	CanPurchase       bool      `json:"can_purchase"`
	CreatedAt         time.Time `json:"created_at"`
	IsUnlocked        bool      `json:"is_unlocked"`
}

type UserCharacter struct {
	UserID         uuid.UUID         `json:"user_id"`
	EquippedHair   *string           `json:"equipped_hair,omitempty"`
	EquippedHat    *string           `json:"equipped_hat,omitempty"`
	EquippedOutfit *string           `json:"equipped_outfit,omitempty"`
	EquippedAura   *string           `json:"equipped_aura,omitempty"`
	EquippedItems  map[string]string `json:"equipped_items,omitempty"`
	UpdatedAt      time.Time         `json:"updated_at"`
}

type CharacterSummary struct {
	Character *UserCharacter   `json:"character"`
	Inventory []*CharacterItem `json:"inventory"`
	Gold      int              `json:"gold_balance"`
}

type EquipItemRequest struct {
	ItemCode string `json:"item_code" binding:"required"`
}

type EquipItemResponse struct {
	Character *UserCharacter `json:"character"`
}

type PurchaseItemRequest struct {
	ItemCode string `json:"item_code" binding:"required"`
}

type PurchaseItemResponse struct {
	Character *UserCharacter `json:"character"`
	Gold      int            `json:"gold_balance"`
	Item      *CharacterItem `json:"item"`
}
